database/redis/notification.go

```diff
@@ -98,7 +134,123 @@ func (connector *DbConnector) removeNotifications(notifications []*moira.Schedul
 }
 
 // FetchNotifications fetch notifications by given timestamp and delete it
-func (connector *DbConnector) FetchNotifications(to int64) ([]*moira.ScheduledNotification, error) {
+func (connector *DbConnector) FetchNotifications(to int64, limit int64) ([]*moira.ScheduledNotification, error) {
+
+	if limit == 0 {
+		return nil, fmt.Errorf("limit musn't be 0")
+	}
+
+	// No limit
+	if limit == notifier.NotificationsLimitUnlimited {
+		return connector.fetchNotificationsNoLimit(to)
+	}
+
+	count, err := connector.notificationsCount(to)
+	if err != nil {
+		return nil, err
+	}
+
+	// Hope count will be not greater then limit when we call fetchNotificationsNoLimit
+	if limit > transactionHeuristicLimit && count < limit/2 {
+		return connector.fetchNotificationsNoLimit(to)
+	}
+
+	return connector.fetchNotificationsWithLimit(to, limit)
+}
+
+func (connector *DbConnector) notificationsCount(to int64) (int64, error) {
+	c := connector.pool.Get()
+	defer c.Close()
+
+	count, err := redis.Int64(c.Do("ZCOUNT", notifierNotificationsKey, "-inf", to))
+
+	if err != nil {
+		return 0, fmt.Errorf("failed to ZCOUNT to notificationsCount: %s", err)
+	}
+
+	return count, nil
+}
+
+// fetchNotificationsWithLimit reads and drops notifications from DB with limit
+func (connector *DbConnector) fetchNotificationsWithLimit(to int64, limit int64) ([]*moira.ScheduledNotification, error) {
+	// fetchNotifecationsWithLimitDo uses WATCH, so transaction may fail and will retry it
+	// see https://redis.io/topics/transactions
+	var e *transactionError
+
+	for i := 0; i < transactionTriesLimit; i++ {
+		res, err := connector.fetchNotificationsWithLimitDo(to, limit)
+
+		if err == nil {
+			return res, nil
+		}
+
+		if !errors.As(err, &e) {
```

Comment:
I can see the root of the problem here. You are implementing an `error` interface for `*transactionError`. You need to change `Error` function receiver from pointer to concrete type and you will be able to make check like in this example:

https://play.golang.org/p/WgOafx0wdsg

Reply:
 Oh, I was stupid. Thank you. Now it works.